### PR TITLE
Support overriding StepUp EntityId

### PR DESCRIPTION
**Pull request by MKodde**
Currently the SAML EntityID of the engineblock SP that us used to do Stepup (SFO) authenticatons to the Stepup-Gateway is always https://<engineblock.sever.domain.name>/authentication/stepup/metadata

The key used to sign the SAML AuthnRequests from this SP is the engineblock default key.

To facilitate a rolling configuration update I want the SP entityID that is used for Stepup to be configurable so that at the same time that the engineblock default key is updated, this entityID can be changed. This then allows two entities, with two different keys, to be configured in the Stepup-Gateway.

https://www.pivotaltracker.com/story/show/186200738

### CODE REVIEW

File: UPGRADING.md [hunk 1]
```diff
@@ -1,5 +1,24 @@
 # UPGRADE NOTES
 
+## 6.13 -> 6.14
+Previously the SAML EntityID of the EngineBlock SP that us used to do Stepup (SFO) authenticatons to the Stepup-Gateway
```
Comment (tvdijen) [inline UPGRADING.md:4]: `was used` and `authentications`

Comment (MKodde) [reply UPGRADING.md]: Thanks for the feedback! Updated the notes according to your typo-suggestions.

File: library/EngineBlock/Corto/ProxyServer.php [hunk 2]
```diff
@@ -487,6 +488,32 @@ public function sendStepupAuthenticationRequest(
...
+        $isEnabled = $features->isEnabled('eb.stepup.sfo.override_engine_entityid');
+
+        if ($isEnabled && $isConfigured) {
+            if (empty($entityIdOverrideValue)) {
+                throw new MissingParameterException(
+                    'When feature "feature_stepup_sfo_override_engine_entityid" is enabled,
+                     you must provide the "stepup.sfo.override_engine_entityid" parameter.'
+                );
+            }
+            $this->_logger->notice(
```
Comment (thijskh) [inline library/EngineBlock/Corto/ProxyServer.php:505]: I think I prefer `info` or even `debug`. I would expect this config to last for an undetermined amount of time, does not make sense to issue notices on every stepup callout about it.

**Review by thijskh — APPROVED** [review]

File: app/config/parameters.yml.dist [hunk 3]
```diff
@@ -261,6 +262,7 @@ parameters:
     stepup.gateway.sfo.sso_location: 'https://gateway.stepup.vm.openconext.org/second-factor-only/single-sign-on'
     ## The public key from the Stepup Gateway IdP
     stepup.gateway.sfo.key_file: /etc/openconext/engineblock.crt
+    stepup.sfo.override_engine_entityid: 'https://engine.vm.openconext.com/new/stepup/metadata'
```
Comment (thijskh) [inline app/config/parameters.yml.dist:265]: The docs/stepup_callout.md file refers to this section for explanation of the Engineblock config. So I think this new setting should have at least a `#` comment like the ones above.

File: library/EngineBlock/Corto/Module/Service/StepupAssertionConsumer.php [hunk 4]
```diff
@@ -138,6 +141,30 @@ public function serve($serviceName, Request $httpRequest)
...
+        $application = EngineBlock_ApplicationSingleton::getInstance();
+        $container = $application->getDiContainer();
+        $entityIdOverrideValue = $container->getStepupEntityIdOverrideValue();
+        $features = $container->getFeatureConfiguration();
+        $isConfigured = $features->hasFeature('eb.stepup.sfo.override_engine_entityid');
+        $isEnabled = $features->isEnabled('eb.stepup.sfo.override_engine_entityid');
+
+        $serviceEntityId = $this->_server->getUrl('stepupMetadataService');
+        if ($isEnabled && $isConfigured) {
+            return $entityIdOverrideValue;
```
Comment (MKodde) [inline library/EngineBlock/Corto/Module/Service/StepupAssertionConsumer.php:163]: Setting the override causes this error down the line when EB is processing the assertion (stepupAssertionConsumerService)

`Response processing failed: "Invalid Assertion in SAML Response, errors: "The configured Service Provider [https://engine.vm.openconext.com/new/stepup/metadata] is not a valid audience for the assertion. Audiences: [https://engine.vm.openconext.org/authentication/stepup/metadata]"`

**Pull request merged**